packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +773,285 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+          if (
+            useForce === false &&
+            (namespaces.length > 1 || namespaces.includes(ALL_NAMESPACES_STRING))
+          ) {
+            return {
+              tag: 'Left',
+              value: {
+                success: false,
+                id,
+                type,
+                error: errorContent(
+                  SavedObjectsErrorHelpers.createBadRequestError(
+                    `Unable to delete saved object that exists in multiple namespaces, use the "force" option to delete it anyway`
+                  )
+                ),
+              },
+            };
+          }
+        }
+        // contains all objects that passed initial preflight checks, including single namespace objects that skipped the mget call
+        // single namespace objects will have namespaces:undefined
+        const expectedResult = {
+          type,
+          id,
+          namespaces,
+          esRequestIndex: indexCounter++,
+        };
+
+        return { tag: 'Right', value: expectedResult };
+      });
+    return expectedBulkDeleteMultiNamespaceDocsResults;
+  }
+
+  /**
+   *  {@inheritDoc ISavedObjectsRepository.bulkDelete}
+   */
+  async bulkDelete(
+    objects: SavedObjectsBulkDeleteObject[],
+    options: SavedObjectsBulkDeleteOptions = {}
+  ): Promise<SavedObjectsBulkDeleteResponse> {
+    const { refresh = DEFAULT_REFRESH_SETTING, force } = options;
+    const namespace = normalizeNamespace(options.namespace);
+    const expectedBulkGetResults = this.presortObjectsByNamespaceType(objects);
+    const multiNamespaceDocsResponse = await this.preflightCheckForBulkDelete({
+      expectedBulkGetResults,
+      namespace,
+    });
+    const bulkDeleteParams: BulkDeleteParams[] = [];
+
+    const expectedBulkDeleteMultiNamespaceDocsResults =
+      this.getExpectedBulkDeleteMultiNamespaceDocsResults({
+        expectedBulkGetResults,
+        multiNamespaceDocsResponse,
+        namespace,
+        force,
+      });
+    // bulk up the bulkDeleteParams
+    expectedBulkDeleteMultiNamespaceDocsResults.map((expectedResult) => {
+      if (isRight(expectedResult)) {
+        bulkDeleteParams.push({
+          delete: {
+            _id: this._serializer.generateRawId(
+              namespace,
+              expectedResult.value.type,
+              expectedResult.value.id
+            ),
+            _index: this.getIndexForType(expectedResult.value.type),
+            ...getExpectedVersionProperties(undefined),
+          },
+        });
+      }
+    });
+
+    const bulkDeleteResponse = bulkDeleteParams.length
+      ? await this.client.bulk({
+          refresh,
+          body: bulkDeleteParams,
+          require_alias: true,
+        })
+      : undefined;
+
+    // extracted to ensure consistency in the error results returned
+    let errorResult: BulkDeleteItemErrorResult;
+    const savedObjects = await Promise.all(
```

Comment:
`deleteLegacyUrlAliases` is asynchronous and only accepts one object at a time, hence having to use `Promise.all` here. We could, potentially, follow a similar strategy to how the `bulkDeleteParams` are created and postpone the call to delete legacy URL aliases before returning the final `statuses` response.

We're not too concerned with the final result from all the calls to `deleteLegacyUrlAliases` so would we still need to wait for the promises to resolve or use Promise.allSettled instead?